### PR TITLE
fix(parser): keep the source exclusion in "tap another untapped" costs (#7522)

### LINKED ISSUE #7522: parser: "Tap another ..." in a TapCreatures cost drops the self-exclusion — 8 cards can pay with the source itself
**Card(s):** Spire Mechcycle (hit in play); also Kumena, Tyrant of Orazca · Meanders Guide · Radiant, Serra Archangel · Shadow Stinger · Sure-Footed Infiltrator · Tyvar, the Pummeler · Veteran Warleader

**Game mode:** AI

## Actual behavior

Spire Mechcycle's exhaust cost reads "Tap **another** untapped Mount or Vehicle you control", but the parsed cost carries no self-exclusion, so the Mechcycle itself is an eligible payment for its own ability.

Parsed cost, taken from a live game state (object 39):

```json
{"type":"TapCreatures","requirement":{"requirement":"count","count":1},
 "filter":{"type":"Or","filters":[
   {"type":"Typed","type_filters":[{"Subtype":"Mount"}],"controller":"You","properties":[]},
   {"type":"Typed","type_filters":[{"Subtype":"Vehicle"}],"controller":"You","properties":[]}]}}
```

`properties` is empty on both branches, and Spire Mechcycle is a Vehicle you control, so it matches its own cost filter.

## Expected behavior

"Another" should reach the cost filter, so the source is not an eligible payment.

The runtime is not the bug. `has_enough_tap_creatures` (`crates/engine/src/game/cost_payability.rs:752`) skips the source only when `exclude_source` is set, and `cost_payability.rs:1112` documents that deliberately — *"CR 601.2b: Standalone TapCreatures (no `{T}`) includes the source itself"*. That is correct for a plain "Tap N untapped creatures you control" cost. `FilterProp::Another` already exists for exactly this exclusion (3200 occurrences in `card-data.json`).

## Class measured

Over all 35,795 cards in `card-data.json`: 150 cards carry a `TapCreatures` ability cost; 8 of them say "another"/"other" in the oracle text; **0 of the 8** carry `FilterProp::Another` in the parsed cost filter.

| Card | abilities affected |
|---|---|
| Kumena, Tyrant of Orazca | 3 |
| Meanders Guide | 1 (triggered) |
| Radiant, Serra Archangel | 1 |
| Shadow Stinger | 1 |
| Spire Mechcycle | 1 |
| Sure-Footed Infiltrator | 1 |
| Tyvar, the Pummeler | 1 |
| Veteran Warleader | 1 |

Each one matches its own cost filter (Kumena is a Merfolk, Radiant has flying, Spire Mechcycle is a Vehicle, and so on), so all 8 can pay with the source.

Crew and Saddle use the same "other" wording, but they are keywords with their own handling — this report is only about the parsed `TapCreatures` ability cost.

## Not verified

I confirmed the parsed data and read the payability code; I did not play out all 8 cards to watch the picker offer the source. Spire Mechcycle is the one I hit in an actual game. The measurement ran on a local `card-data.json` dated 2026-08-12; the Spire Mechcycle cost in a live 2026-08-17 game state is identical.

### PR DESCRIPTION
**Pull request by cuinhellcat**
Closes #7522.

## Defect

`parse_oracle_cost`'s tap-cost branch consumed `"another untapped "` in the same `alt` as `"an untapped "` / `"an "`, all yielding a bare count of 1. The qualifier was gone from the phrase handed to `parse_target`, so the cost filter came out without `FilterProp::Another` and the ability source was an eligible payment for its own cost. Spire Mechcycle is a Vehicle paying *"Tap another untapped Mount or Vehicle you control"*.

## Fix

The leading-quantifier `alt` now reports the typed `CountWord` that the sacrifice imperative already uses for this exact failure (#4513), and the exclusion is applied to every leg of the built filter. `apply_source_exclusion` is extracted from the existing `ensure_another_sacrifice_filter` so both cost paths share one authority.

The numeric branch (*"tap two other untapped artifacts you control"*) is unchanged: it never consumes `"other"`, so the phrase reaches `parse_type_phrase` intact and that supplies the property. A test row pins it.

The runtime is not touched. `has_enough_tap_creatures` evaluates the cost filter against a `FilterContext::from_source`, so the property takes effect the moment the parser emits it; the separate `exclude_source` flag for composite `{T}` costs is untouched.

## Class

Measured over the 96 distinct printed tap-cost phrases in `client/public/card-data.json` (35,795 cards):

| | before | after |
|---|---|---|
| phrases parsing to `TapCreatures` | 94 | 94 |
| of those, saying "another"/"other" | 11 | 11 |
| of those, carrying `FilterProp::Another` | 4 | 11 |

The 7 newly fixed phrases cover 11 cards / 11 activated abilities: Black Oak of Odunos, Kumena Tyrant of Orazca, Network Terminal, Radiant Serra Archangel, Ranger's Hawk, Shadow Stinger, Spire Mechcycle, Sure-Footed Infiltrator, Tyvar the Pummeler, Veteran Warleader, Wanderbrine Trapper.

Meanders Guide's triggered *"you may tap another untapped Merfolk you control"* runs through the same branch and is fixed too — measured separately through `parse_oracle_text`: filter `properties` `[]` before, `[Another]` after. **12 cards in total.** The issue reported 8; the measured class is 12. I will correct the issue.

## Counter-probe

With `CountWord::SourceExclusion => filter`: `tap_cost_another_carries_the_source_exclusion` fails with `"Tap another untapped Merfolk you control" exclusion mismatch, got []`, `tap_cost_another_marks_every_leg_of_a_disjunction` fails, and `the_source_alone_cannot_pay_its_own_tap_another_cost` fails. The two counter-direction tests (`a_second_untapped_creature_pays_the_tap_another_cost`, `a_plain_tap_cost_still_includes_the_source`) stay green — the second is also the reach guard that keeps the first test's negative assertion from being vacuous.

## Not covered

The "other than this creature" tail forms reach the exclusion through a different grammar and were already correct (Impelled Giant). Mossbridge Troll's *"tap any number of untapped creatures you control other than this creature with total power 10 or greater"* does not lower to a `TapCreatures` cost at all — it becomes an `EffectCost` — and this change does not address that.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected parsing of “tap another” and similar tap-cost restrictions.
  * Preserved source exclusions across alternative and combined tap-cost conditions.
  * Prevented an ability’s source from incorrectly satisfying its own “another” requirement.
  * Preserved normal behavior for unrestricted tap costs.

* **Tests**
  * Added coverage for legal actions, activation rejection, successful payment, and multi-option tap-cost conditions.

### CODE REVIEW

**Comment by matthewevans** [comment]
Maintainer fixup is now at `279c2beb05f6333b5c329c39a3c264b162d64ccf`:

- reused `oracle_target::distribute_shared_properties` so source exclusion reaches typed legs through `Or` and `And` without changing the negated ownership predicate;
- added the conjunctive parser regression and a real `activate(...).pay_with(...).resolve()` payment assertion; and
- corrected the cost annotations against CR 602.2b and CR 118.3.

The previous CodeRabbit review and parse-diff sticky are for `2ff39f9`, so this head is held for fresh CI, a current-head parse-diff sticky, and current-head independent review feedback before enqueue.

**Comment by matthewevans** [comment]
**Maintainer hold — current-head evidence refresh pending.** I pushed a scoped review fixup at `2cc86406c7521efabdce42459760cdff5ca45d04`: the lone-source regression now proves ability 0 was published as a concrete `TapCreatures` cost with no `Effect::Unimplemented`, and its activation-cost annotation now cites CR 602.2b + CR 601.2h + CR 118.3 rather than claiming CR 601.2b establishes source eligibility.

The `bug` label remains. This PR is awaiting fresh CI and current-head parser/coverage evidence before the review can resume; no approval or enqueue is implied by this hold.

**Comment by cuinhellcat** [comment]
Thanks — both fixups are strictly better than what I wrote, and both were my misses.

`oracle_target::distribute_shared_properties` already answered "put these properties on every typed leg of this filter tree", and I wrote a second one anyway; mine also only recursed through `Or`, so the `And` shape you added a regression for would have silently dropped the exclusion. And the annotations were wrong in the same direction as the last round: CR 109.4 / CR 701.21a describe the sacrifice case, not an activation cost.

Nothing outstanding from me on this branch — I have not pushed since `2ff39f9` and will not, so the head stays where you put it.

File: crates/engine/tests/integration/tap_cost_another_self_exclusion.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+/// payable and the ability is offered. Guards against over-suppression, the
+/// expensive collateral of a self-exclusion fix.
+#[test]
+fn a_second_untapped_creature_pays_the_tap_another_cost() {
+    let (runner, source) = board(ANOTHER, 1);
+    assert!(
+        offers_activation(&runner, source),
+        "another untapped creature makes the cost payable"
+    );
+}
```
[the remark this answers is not in the document]
Comment (matthewevans) [reply crates/engine/tests/integration/tap_cost_another_self_exclusion.rs:130]: Addressed on the current head `2cc86406c7521efabdce42459760cdff5ca45d04`: `a_second_untapped_creature_pays_the_tap_another_cost` now executes `runner.activate(source, 0).pay_with(&[helper]).resolve()` and asserts that the helper, rather than the source, was tapped.

**Review by matthewevans — APPROVED** [review]
Approved for merge queue: the current head preserves source exclusion through the parser filter tree, and the regression executes the real activation payment path with a distinct helper.
